### ISSUE TITLE
Deadlock on Main Thread when compiling with 3.1.8

### ISSUE THREAD

**Issue by pmlt** [emscripten-core/emscripten#16667]
Hi,

Yesterday we tried upgrading the toolchain to 3.1.8, and now we're getting a deadlock when calling `fopen` from the main thread. It does not happen the first time, only after 2-3 file reads. Each file is closed before the next one is opened.

Here's the callstack of the main thread when the deadlock is in effect:

![callstack](https://user-images.githubusercontent.com/1267967/162198575-eeb7ac04-5ee3-4473-9569-7ac33414ab93.png)

If I revert back to 3.1.7 (using `emsdk install 3.1.7 && emsdk activate 3.1.7`), and do a clean build, the same code does not deadlock, so it looks like a real regression.

Unfortunately I don't have a simple repro I can share, but I can provide more information if someone tells me what to look for!

**Comment by kripken** [comment]
I think the best strategy here might be bisection between those versions. See docs here, it's fairly simple to do (you don't need to compile from source or anything):

https://emscripten.org/docs/contributing/developers_guide.html#bisecting

That should get you to a specific commit (or hopefully a short list of commits).

**Comment by pmlt** [comment]
Hi @kripken,

Sorry for the delay. We have followed the procedure and identified commit dbc650f0b6b6cc7e70329017ba4984db1d476f67 as the first bad commit from the emscripten-releases repository. So if I got this right, it corresponds to this change:

https://chromium.googlesource.com/external/github.com/emscripten-core/emscripten.git/+log/1a0b77c572ad..c48f73a5c763

Does that make sense?

**Comment by kripken** [comment]
Looks like #16339 then. Yes, that makes sense, so I think you got it right.

cc @sbc100

**Comment by sbc100** [comment]
@pmlt, can you include the full stacktrace rather than just tip of it?   Does building with `-sEXIT_RUNTIME` make the problem go away?

**Comment by pmlt** [comment]
Hi @sbc100,

I can confirm that linking with -sEXIT_RUNTIME using 3.1.8 fixes the deadlock. Here's the full callstack of the deadlock:

```
a_cas_p (atomic_arch.h:36)
futex_wait_busy (emscripten_futex_wait.c:43)
emscripten_futex_wait (emscripten_futex_wait.c:125)
__wait (__wait.c:32)
__futexwait (pthread_impl.h:201)
__lock (__lock.c:44)
__ofl_lock (ofl.c:11)
__ofl_add (ofl_add.c:5)
__fdopen (__fdopen.c:60)
fopen (fopen.c:28)
CAkFileHelpers::OpenFile(char const*, AkOpenMode, bool, bool, _IO_FILE*&) (AkFileHelpers.h:88)
CAkFileHelpers::Open(char const*, AkOpenMode, bool, AkFileDesc&) (AkFileHelpers.h:107)
CAkMultipleFileLocation::Open(char const*, AkOpenMode, AkFileSystemFlags*, bool, AkFileDesc&) (AkMultipleFileLocation.cpp:99)
CAkDefaultIOHookBlocking::Open(char const*, AkOpenMode, AkFileSystemFlags*, bool&, AkFileDesc&) (AkDefaultIOHookBlocking.cpp:120)
CAkFilePackageLowLevelIO<CAkDefaultIOHookBlocking, CAkDiskPackage>::Open(char const*, AkOpenMode, AkFileSystemFlags*, bool&, AkFileDesc&) (AkFilePackageLowLevelIO.inl:102)
AK::StreamMgr::CAkStreamMgr::CreateStd(char const*, AkFileSystemFlags*, AkOpenMode, AK::IAkStdStream*&, bool) (AkStreamMgr.cpp:636)
CAkBankReader::SetFile(char const*, unsigned int, unsigned int, unsigned int, void*) (AkBankReader.cpp:133)
CAkBankMgr::SetFileReader(unsigned int, char const*, unsigned int, unsigned int, void*, bool) (AkBankMgr.cpp:1126)
CAkBankMgr::LoadBank(CAkBankMgr::AkBankQueueItem, CAkUsageSlot*&, CAkBankMgr::AkLoadBankDataMode, bool, bool) (AkBankMgr.cpp:1263)
CAkBankMgr::LoadBankPre(CAkBankMgr::AkBankQueueItem&) (AkBankMgr.cpp:1465)
CAkBankMgr::ExecuteCommand(CAkBankMgr::AkBankQueueItem&) (AkBankMgr.cpp:485)
CAkBankMgr::QueueBankCommand(CAkBankMgr::AkBankQueueItem) (AkBankMgr.h:460)
AK::SoundEngine::LoadBankInternal(unsigned int, unsigned int, AkFileNameString, AkBankLoadFlag, CAkBankMgr::AkBankQueueItemType, void (*)(unsigned int, void const*, AKRESULT, void*), void*, void const*, unsigned int) (AkAudioLib.cpp:4224)
AK::SoundEngine::LoadBank(char const*, unsigned int&, unsigned int) (AkAudioLib.cpp:4449)
DemoLocalization::Init() (DemoLocalization.cpp:46)
Menu::StackPage(Page*) (Menu.cpp:146)
DemoDialogueMenu::LocalizationButton_Pressed(void*, ControlEvent*) (DemoDialogueMenu.cpp:63)
Control::CallDelegate(ControlEvent*) const (Control.cpp:47)
ButtonControl::Update(UniversalGamepad const&) (ButtonControl.cpp:40)
Page::Update() (Page.cpp:121)
Menu::Update() (Menu.cpp:76)
IntegrationDemo::Update() (IntegrationDemo.cpp:187)
main_loop_iteration() (Main.cpp:45)
callUserCallback (IntegrationDemo.js:5690)
runIter (IntegrationDemo.js:5745)
Browser_mainLoop_runner (IntegrationDemo.js:5650)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
requestAnimationFrame (async)
requestAnimationFrame (IntegrationDemo.js:6073)
Browser_mainLoop_scheduler_rAF (IntegrationDemo.js:5526)
Browser_mainLoop_runner (IntegrationDemo.js:5663)
```

**Comment by sbc100** [comment]
In the working build, if you build with `-sASSERTIONS` do you see any messages in your console?  In particular, do you see "user callback triggered after runtime exited or application aborted"?

**Comment by sbc100** [comment]
Does your application have any ability to shutdown or exit, or does it just run forever (i,e. for the lifetime of the tab its in)?    Does is join all its worker threads in this case?

**Comment by Username404-59** [comment]
This may be a duplicate of #16654

**Comment by sbc100** [comment]
You can try building with `-sPTHREADS_DEBUG` to get some more logging, and perhaps we need to add some logging around the open file lock too.

**Comment by pmlt** [comment]
@sbc100, 

I didn't see anything useful in the log with PTHREADS_DEBUG, but I've attached it to this comment anyway:
[deadlock-console.log](https://github.com/emscripten-core/emscripten/files/8484219/deadlock-console.log)

We are not opening files from other threads. I compiled with `-sSAFE_HEAP` and caught a memory alignment violation in our code, but fixing it did not fix the deadlock. I tried compiling with ASAN and UBSAN, but I'm getting linker errors so that will have to be for another day.

We do have a custom memory allocator that can detect stomps for our Linux port using `mprotect`. I will try to port it to Emscripten and see if something comes up.

**Comment by aharondavid** [comment]
I think that I know what cause for that, the function  `__stdio_exit(void)` (in file: c:\emsdk\upstream\emscripten\system\lib\libc\musl\src\stdio\__stdio_exit.c)
is used in `__ofl_lock ` without  `__ofl_unlock`

**Comment by sbc100** [comment]
Correct, that is what `__stdio_exit` does, and the incorrect use of `__stdio_exit` was fixed in #17046 which was released in 3.1.12.   
So I think this issue can now be closed?    @pmlt can you confirm?

**Comment by pmlt** [comment]
@sbc100 I can confirm that revision 3.1.12 and 3.1.17 compile and run correctly. The deadlock is fixed!
